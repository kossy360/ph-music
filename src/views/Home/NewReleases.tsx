import dashIcon from '@iconify-icons/bi/dash';
import plusIcon from '@iconify-icons/bi/plus';
import Icon from '@iconify/react';
import React, { useMemo, useState } from 'react';
import { useInfiniteQuery } from 'react-query';
import { useSelector } from 'react-redux';
import styled, { useTheme } from 'styled-components';
import DataLoader from '../../components/DataLoader';
import LoadMoreButton from '../../components/LoadMoreButton';
import { PlainButton } from '../../components/PlainButton';
import { Typography } from '../../components/Typography';
import useNotification from '../../hooks/useNotification';
import useSaveToLibrary from '../../hooks/useSaveToLibrary';
import useScreenSize from '../../hooks/useScreenSize';
import { newReleasesService } from '../../services/spotifyService';

const Container = styled.div`
  padding: 1rem;
  width: 90%;
  max-width: 80rem;
  margin: auto;

  .albumsGrid {
    padding: 2rem 0;
    margin-top: 10px;
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(15rem, 1fr));
    grid-template-rows: auto;
    column-gap: 1rem;
    row-gap: 2rem;

    .albumCard {
      display: grid;
      justify-content: center;
      gap: 5px;

      .albumArt {
        display: grid;
        justify-content: center;

        img {
          width: 15rem;
        }
      }
    }
  }
`;

interface IProps {
  isSearching: boolean;
}

const NewReleases = (props: IProps) => {
  const theme = useTheme();
  const sw = useScreenSize();
  const notify = useNotification();
  const savedAlbums = useSelector((state) => state.library.albums);
  const library = useSaveToLibrary('albums');
  const [limit] = useState(20);
  const newReleases = useInfiniteQuery(
    ['search', 'new-releases', limit],
    ({ pageParam = 0 }) => newReleasesService(pageParam, limit),
    {
      getNextPageParam: (lp) => (lp.albums.next ? lp.albums.offset + 1 : false),
      onError: () => notify.error('Error fetching new releases'),
    }
  );
  const items = useMemo(() => {
    return newReleases.data?.pages.map((p) => p.albums.items).flat() ?? [];
  }, [newReleases.data]);
  const itemCount = props.isSearching ? 4 : items.length ?? 0;

  if (sw.max(1200) && props.isSearching) return null;

  return (
    <Container>
      <Typography as="h1" textStyle="sm18">
        New Releases
      </Typography>
      <div role="list" aria-label="album list" className="albumsGrid">
        {items.slice(0, itemCount).map((album) => {
          const isSaved = savedAlbums[album.id] ?? false;

          return (
            <div key={album.id} role="listitem" className="albumCard">
              <div className="albumArt">
                <img
                  src={
                    album.images[0]?.url ?? 'https://picsum.photos/id/237/200'
                  }
                  alt="album art"
                />
              </div>
              <Typography
                textStyle="sm16"
                textAlign="center"
                textTheme={{ weight: 500 }}
                ellipsis
              >
                {album.name}
              </Typography>
              <PlainButton onClick={() => library.toggleSave(album.id)}>
                <Icon
                  icon={isSaved ? dashIcon : plusIcon}
                  height={'1.3rem'}
                  color={theme.text.colors.primary}
                />
                <Typography>{` ${
                  isSaved ? 'Remove from library' : 'Add to library'
                }`}</Typography>
              </PlainButton>
            </div>
          );
        })}
      </div>
      <DataLoader
        isLoading={newReleases.isLoading}
        hasData={items.length > 0}
        noDataText="Sorry, there is nothing new"
      />
      {!props.isSearching && <LoadMoreButton query={newReleases} />}
    </Container>
  );
};

export default NewReleases;
