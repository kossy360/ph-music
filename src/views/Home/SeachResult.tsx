import dashIcon from '@iconify-icons/bi/dash';
import plusIcon from '@iconify-icons/bi/plus';
import Icon from '@iconify/react';
import React, { useMemo } from 'react';
import { useInfiniteQuery } from 'react-query';
import { useSelector } from 'react-redux';
import styled, { useTheme } from 'styled-components';
import DataLoader from '../../components/DataLoader';
import LoadMoreButton from '../../components/LoadMoreButton';
import { PlainButton } from '../../components/PlainButton';
import { Typography } from '../../components/Typography';
import useNotification from '../../hooks/useNotification';
import useSaveToLibrary from '../../hooks/useSaveToLibrary';
import useScreenSize from '../../hooks/useScreenSize';
import { searchTracksService } from '../../services/spotifyService';

const Container = styled.div`
  width: 90%;
  max-width: 80rem;
  margin: auto;
  padding-bottom: 2rem;
  padding-top: 1rem;

  .tableWrapper {
    padding: 2rem 0;
    width: 100%;
  }

  .resultTable {
    width: 100%;
    border-collapse: collapse;

    .searchResultRow {
      td {
        padding: 10px;
        border-bottom: 1px solid #d9dadc;
      }

      .trackAvatar {
        img {
          width: 2rem;
          border-radius: 50%;
        }
      }
    }
  }

  .resultContainerMobile {
    display: grid;
    gap: 1rem;

    .searchResultItem {
      display: grid;
      grid-template-columns: auto 1fr auto;
      align-items: center;
      gap: 10px;
      border-bottom: 1px solid #d9dadc;

      .trackAvatar {
        img {
          width: 4rem;
          border-radius: 50%;
        }
      }

      .details {
        display: grid;
      }
    }
  }
`;

const msToMinutes = (ms: number) => {
  const minutes = Math.floor(ms / 60000);
  const seconds = Math.floor((ms % 60000) / 1000);

  return `${minutes}:${seconds < 10 ? '0' : ''}${seconds}`;
};

interface IProps {
  search: string;
}

const SearchResult = (props: IProps) => {
  const { search } = props;
  const theme = useTheme();
  const notify = useNotification();
  const sw = useScreenSize();
  const savedTracks = useSelector((state) => state.library.tracks);
  const { toggleSave } = useSaveToLibrary('tracks');
  const searchResults = useInfiniteQuery(
    ['search', 'tracks', search],
    ({ pageParam = 0 }) =>
      searchTracksService({ q: search, offset: pageParam }),
    {
      enabled: !!search,
      getNextPageParam: (lp) => (lp.tracks.next ? lp.tracks.offset + 1 : false),
      onError: () => notify.error('Error fetching search results'),
    }
  );
  const items = useMemo(() => {
    return searchResults.data?.pages.map((p) => p.tracks.items).flat() ?? [];
  }, [searchResults.data]);

  if (!props.search) return null;

  return (
    <Container>
      <Typography as="h1" textStyle="sm18">
        Search Result
      </Typography>
      <div className="tableWrapper">
        {sw.min(651) && (
          <table aria-label="search results" className="resultTable">
            <tbody>
              {items.map((track) => {
                const isSaved = savedTracks[track.id] ?? false;

                return (
                  <tr key={track.id} className="searchResultRow">
                    <td>
                      <div className="trackAvatar">
                        <img
                          src={
                            track.album.images[0]?.url ??
                            'https://picsum.photos/id/237/200'
                          }
                          alt="album art"
                        />
                      </div>
                    </td>
                    <td>
                      <Typography>{track.name}</Typography>
                    </td>
                    <td>
                      <Typography>{track.album.name}</Typography>
                    </td>
                    <td>
                      <Typography>{msToMinutes(track.duration_ms)}</Typography>
                    </td>
                    <td>
                      <PlainButton onClick={() => toggleSave(track.id)}>
                        <Icon
                          icon={isSaved ? dashIcon : plusIcon}
                          height={'1.3rem'}
                          color={theme.text.colors.primary}
                        />
                      </PlainButton>
                    </td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        )}
        {sw.max(650) && (
          <div
            role="list"
            aria-label="search results"
            className="resultContainerMobile"
          >
            {items.map((track) => {
              const isSaved = savedTracks[track.id] ?? false;

              return (
                <div
                  key={track.id}
                  role="listitem"
                  className="searchResultItem"
                >
                  <div className="trackAvatar">
                    <img
                      src={
                        track.album.images[0]?.url ??
                        'https://picsum.photos/id/237/200'
                      }
                      alt="album art"
                    />
                  </div>
                  <div className="details">
                    <Typography ellipsis>{track.name}</Typography>
                    <Typography ellipsis>{track.album.name}</Typography>
                    <Typography>{msToMinutes(track.duration_ms)}</Typography>
                  </div>
                  <PlainButton onClick={() => toggleSave(track.id)}>
                    <Icon
                      icon={isSaved ? dashIcon : plusIcon}
                      height={'1.3rem'}
                      color={theme.text.colors.primary}
                    />
                  </PlainButton>
                </div>
              );
            })}
          </div>
        )}
      </div>
      <DataLoader
        isLoading={searchResults.isLoading}
        hasData={items.length > 0}
        noDataText="Sorry, no matching tracks were found"
      />
      <LoadMoreButton query={searchResults} />
    </Container>
  );
};

export default SearchResult;
