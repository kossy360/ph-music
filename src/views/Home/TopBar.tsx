import menuIcon from '@iconify-icons/feather/menu';
import xIcon from '@iconify-icons/feather/x';
import Icon from '@iconify/react';
import { Drawer } from '@material-ui/core';
import { throttle } from 'lodash';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';
import styled, { useTheme } from 'styled-components';
import LogoutButton from '../../components/LogoutButton';
import LogoutButtonText from '../../components/LogoutButtonText';
import { PlainButton } from '../../components/PlainButton';
import { Typography } from '../../components/Typography';
import UserAvatar from '../../components/UserAvatar';
import useScreenSize from '../../hooks/useScreenSize';

const Container = styled.div`
  border-bottom: 1px solid #d9dadc;
  width: 100%;
  padding: 5px 10px;

  .searchInput {
    text-align: center;
    font-family: 'fira sans';
    flex-grow: 1;
    padding: 6px;
    max-width: 30rem;
    background-color: #f3f3f3;
    border: none;
    border-radius: 5px;
    outline: none;
  }

  .topBar {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 7px 14px;

    .libraryLogout {
      display: grid;
      grid-auto-flow: column;
      align-items: center;
      gap: 2rem;

      .logoutButton {
        padding: 0;
        display: flex;
        align-items: center;
        border: 2px solid;
        background: none;
        padding: 0 20px;
        border-radius: 5px;
        cursor: pointer;
      }
    }
  }

  .topBarMobile {
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: start;
    gap: 10px;

    .inputWrapper {
      display: flex;
      justify-content: center;
    }
  }
`;

const StyledDrawer = styled(Drawer)`
  width: 70%;

  .MuiPaper-root {
    width: 70%;
    max-width: 45rem;
  }

  .menuHeader {
    display: flex;
    justify-content: flex-end;
    width: 100%;
    align-items: center;
    padding: 1rem;
  }

  .menuContainer {
    width: 100%;
    padding: 0 1rem;
    background-color: #fff;
    display: grid;
    grid-auto-flow: row;
    gap: 1rem;
    justify-content: start;
  }
`;

interface IProps {
  onSearchChange: (val: string) => void;
}

const TopBar = (props: IProps) => {
  const { onSearchChange } = props;
  const location = useLocation();
  const theme = useTheme();
  const [open, setOpen] = useState(false);
  const screenWidth = useScreenSize();
  const [search, setSearch] = useState('');

  const throttledSearch = useMemo(
    () => throttle(onSearchChange, 600, { trailing: true }),
    [onSearchChange]
  );
  const inputRef = useRef<HTMLInputElement | null>(null);

  useEffect(() => {
    if ((location.state as any)?.focusInput) {
      inputRef.current?.focus();
    }
  }, [location.state]);

  useEffect(() => {
    throttledSearch(search);
  }, [search, throttledSearch]);

  return (
    <Container>
      {screenWidth.min(861) && (
        <div className="topBar">
          <UserAvatar />
          <input
            ref={inputRef}
            type="text"
            aria-label="search field"
            className="searchInput"
            placeholder="search"
            value={search}
            onChange={(e) => setSearch(e.target.value)}
          />
          <div className="libraryLogout">
            <Link to="/library">
              <Typography textStyle="sm18">My Library</Typography>
            </Link>
            <LogoutButton />
          </div>
        </div>
      )}
      {screenWidth.max(860) && (
        <div className="topBarMobile">
          <PlainButton aria-label="open menu" onClick={() => setOpen(true)}>
            <Icon
              icon={menuIcon}
              height={'2rem'}
              color={theme.text.colors.primary}
            />
          </PlainButton>
          <div className="inputWrapper">
            <input
              ref={inputRef}
              type="text"
              aria-label="search field"
              className="searchInput"
              placeholder="search"
              value={search}
              onChange={(e) => setSearch(e.target.value)}
            />
          </div>
          <StyledDrawer open={open} onClose={() => setOpen(false)}>
            <div
              role="navigation"
              aria-label="app drawer"
              className="menuHeader"
            >
              <PlainButton
                aria-label="close menu"
                onClick={() => setOpen(false)}
              >
                <Icon
                  icon={xIcon}
                  height={'2rem'}
                  color={theme.text.colors.primary}
                />
              </PlainButton>
            </div>
            <div className="menuContainer">
              <UserAvatar />
              <Link to="/library">
                <Typography textStyle="sm16">My Library</Typography>
              </Link>
              <LogoutButtonText />
            </div>
          </StyledDrawer>
        </div>
      )}
    </Container>
  );
};

export default TopBar;
