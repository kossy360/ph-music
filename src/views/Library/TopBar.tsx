import menuIcon from '@iconify-icons/feather/menu';
import xIcon from '@iconify-icons/feather/x';
import Icon from '@iconify/react';
import { Drawer } from '@material-ui/core';
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import styled, { useTheme } from 'styled-components';
import LogoutButton from '../../components/LogoutButton';
import LogoutButtonText from '../../components/LogoutButtonText';
import { PlainButton } from '../../components/PlainButton';
import { Typography } from '../../components/Typography';
import UserAvatar from '../../components/UserAvatar';
import useScreenSize from '../../hooks/useScreenSize';
import ExportForm from './ExportForm';

const Container = styled.div`
  border-bottom: 1px solid #d9dadc;
  width: 100%;
  padding: 7px 14px;

  .searchInput {
    text-align: center;
    font-family: 'fira sans';
    flex-grow: 1;
    padding: 6px;
    max-width: 30rem;
    background-color: #f3f3f3;
    border: none;
    border-radius: 5px;
    outline: none;
  }

  .topBar {
    display: flex;
    justify-content: space-between;
    align-items: center;

    .libraryLogout {
      display: grid;
      grid-auto-flow: column;
      align-items: center;
      gap: 2rem;
    }

    .avatarExport {
      display: grid;
      grid-auto-flow: column;
      gap: 3rem;
    }
  }

  .topBarMobile {
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: start;
    gap: 10px;

    .titleWrapper {
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
`;

const StyledDrawer = styled(Drawer)`
  width: 70%;

  .MuiPaper-root {
    width: 70%;
    max-width: 20rem;
  }

  .menuHeader {
    display: flex;
    justify-content: flex-end;
    width: 100%;
    align-items: center;
    padding: 1rem;
  }

  .menuContainer {
    width: 100%;
    padding: 0 1rem;
    background-color: #fff;
    display: grid;
    grid-auto-flow: row;
    gap: 1rem;
    justify-content: start;
  }
`;

const TopBar = () => {
  const theme = useTheme();
  const [open, setOpen] = useState(false);
  const screenWidth = useScreenSize();
  const navigate = useNavigate();

  return (
    <Container>
      {screenWidth.min(861) && (
        <div className="topBar">
          <div className="avatarExport">
            <UserAvatar />
            <ExportForm />
          </div>
          <Typography textStyle="sm18" textTheme={{ weight: 500 }}>
            My Library
          </Typography>
          <div className="libraryLogout">
            <PlainButton
              onClick={() => {
                navigate('/', { state: { focusInput: true } });
              }}
            >
              <Typography textStyle="sm18">Search</Typography>
            </PlainButton>
            <LogoutButton />
          </div>
        </div>
      )}
      {screenWidth.max(860) && (
        <div className="topBarMobile">
          <PlainButton aria-label="open drawer" onClick={() => setOpen(true)}>
            <Icon
              icon={menuIcon}
              height={'2rem'}
              color={theme.text.colors.primary}
            />
          </PlainButton>
          <div className="titleWrapper">
            <Typography textStyle="sm18" textTheme={{ weight: 500 }}>
              My Library
            </Typography>
          </div>
          <StyledDrawer
            role="navigation"
            aria-label="app drawer"
            open={open}
            onClose={() => setOpen(false)}
          >
            <div className="menuHeader">
              <PlainButton
                aria-label="close menu"
                onClick={() => setOpen(false)}
              >
                <Icon
                  icon={xIcon}
                  height={'2rem'}
                  color={theme.text.colors.primary}
                />
              </PlainButton>
            </div>
            <div className="menuContainer">
              <UserAvatar />
              <ExportForm />
              <PlainButton
                justify="start"
                onClick={() => {
                  navigate('/', { state: { focusInput: true } });
                }}
              >
                <Typography textStyle="sm16">Search</Typography>
              </PlainButton>
              <LogoutButtonText />
            </div>
          </StyledDrawer>
        </div>
      )}
    </Container>
  );
};

export default TopBar;
