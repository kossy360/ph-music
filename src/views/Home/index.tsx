import React, { useState } from 'react';
import styled from 'styled-components';
import NewReleases from './NewReleases';
import SearchResult from './SeachResult';
import TopBar from './TopBar';

const Container = styled.div`
  display: grid;
  grid-auto-rows: max-content 1fr;
  height: 100%;

  .bottomSection {
    height: 100%;
    overflow: auto;
  }
`;

const Home = () => {
  const [search, setSearch] = useState('');

  return (
    <Container>
      <TopBar onSearchChange={setSearch} />
      <div className="bottomSection">
        <NewReleases isSearching={!!search} />
        <SearchResult search={search} />
      </div>
    </Container>
  );
};

export default Home;
