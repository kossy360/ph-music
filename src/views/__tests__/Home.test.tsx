import { useWindowWidth } from '@react-hook/window-size';
import { fireEvent, waitFor, within } from '@testing-library/react';
import { startMirage } from '../../test/mirageServer';
import { authenticatedTestStore } from '../../test/redux';
import { renderWithApp } from '../../test/testRenderer';
import Home from '../Home';
import NewReleases from '../Home/NewReleases';
import SearchResult from '../Home/SeachResult';
import TopBar from '../Home/TopBar';

const setScreenSize = (size: number) => {
  return (
    useWindowWidth as jest.MockedFunction<typeof useWindowWidth>
  ).mockImplementation(() => size);
};

type TMirageServer = ReturnType<typeof startMirage>;
let server: TMirageServer | null = null;

jest.mock('@react-hook/window-size');

beforeEach(() => {
  server = startMirage();
});

afterEach(() => {
  server?.shutdown();
});

describe('Top navigator', () => {
  it('should render properly (desktop)', async () => {
    // set screen width
    (
      useWindowWidth as jest.MockedFunction<typeof useWindowWidth>
    ).mockImplementation(() => 1200);

    const component = renderWithApp(<TopBar onSearchChange={() => {}} />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });

    const avatar = component.getByRole('img', {
      name: /user avatar/i,
    });
    const input = component.getByRole('textbox', {
      name: /search field/i,
    });
    const libraryLink = component.getByText(/my library/i);
    const logoutButton = component.getByRole('button', {
      name: /logout button/i,
    });

    expect(avatar).toBeDefined();
    expect(input).toBeDefined();
    expect(libraryLink).toBeDefined();
    expect(logoutButton).toBeDefined();
  });

  it('should render properly (mobile)', async () => {
    // set screen width
    (
      useWindowWidth as jest.MockedFunction<typeof useWindowWidth>
    ).mockImplementation(() => 400);

    const component = renderWithApp(<TopBar onSearchChange={() => {}} />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const menuButton = component.getByRole('button', {
      name: /open menu/i,
    });

    component.getByRole('textbox', {
      name: /search field/i,
    });
    fireEvent.click(menuButton);

    await waitFor(() =>
      component.getByRole('navigation', { name: /app drawer/! })
    );

    component.getByRole('img', {
      name: /user avatar/i,
    });
    component.getByText(/my library/i);
    component.getByText(/logout/i);

    const closeButton = component.getByRole('button', { name: /close menu/! });

    fireEvent.click(closeButton);

    expect(
      component.queryByRole('navigation', { name: /app drawer/! })
    ).toBeNull();
  });
});

describe('New releases section', () => {
  it('should render all items', async () => {
    const data = server!.createList('spotifyAlbum', 5).map((m) => m.album);
    const component = renderWithApp(<NewReleases isSearching={false} />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const listbox = component.getByRole('list', {
      name: /album list/i,
    });
    const albums = await waitFor(() => component.getAllByRole('listitem'));

    expect(listbox).toBeDefined();
    expect(albums.length).toEqual(data.length);
  });

  it('should render only 4 items when searching', async () => {
    server!.createList('spotifyAlbum', 10).map((m) => m.album);
    const component = renderWithApp(<NewReleases isSearching={true} />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const listbox = component.getByRole('list', {
      name: /album list/i,
    });
    const albums = await waitFor(() => component.getAllByRole('listitem'));

    expect(listbox).toBeDefined();
    expect(albums.length).toEqual(4);
  });

  it('should not render when searching on mobile screens', async () => {
    server!.createList('spotifyAlbum', 10).map((m) => m.album);
    (
      useWindowWidth as jest.MockedFunction<typeof useWindowWidth>
    ).mockImplementation(() => 400);

    const component = renderWithApp(<NewReleases isSearching={true} />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });

    const listbox = component.queryByRole('list', {
      name: /album list/i,
    });

    expect(listbox).toBeNull();
  });
});

describe('Search results section', () => {
  it('should render table on large screens', async () => {
    setScreenSize(1200);
    const data = server!.createList('spotifyTrack', 10).map((m) => m.track);
    const component = renderWithApp(<SearchResult search="search" />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const table = component.getByRole('table');
    const tracks = await waitFor(() => within(table).getAllByRole('row'));
    const firstTrack = tracks[0];

    within(firstTrack).getByText(data[0].name);

    expect(tracks.length).toEqual(data.length);
  });

  it('should render table on mobile screens', async () => {
    setScreenSize(400);
    const data = server!.createList('spotifyTrack', 10).map((m) => m.track);
    const component = renderWithApp(<SearchResult search="search" />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const listbox = component.getByRole('list', {
      name: /search results/i,
    });
    const tracks = await waitFor(() =>
      within(listbox).getAllByRole('listitem')
    );
    const firstTrack = tracks[0];

    within(firstTrack).getAllByText(data[0].name);
    expect(listbox).toBeDefined();
    expect(tracks.length).toEqual(data.length);
  });
});

describe('Home Screen', () => {
  it('should render table on large screens', async () => {
    setScreenSize(1500);
    const trackData = server!
      .createList('spotifyTrack', 10)
      .map((m) => m.track);
    const albumData = server!
      .createList('spotifyAlbum', 10)
      .map((m) => m.album);
    const component = renderWithApp(<Home />, {
      store: authenticatedTestStore(),
      routerProps: { initialEntries: ['/'] },
    });
    const albumSection = component.getByRole('list', { name: /album list/i });
    const albumItems = await waitFor(() => component.getAllByRole('listitem'));

    expect(albumItems.length).toEqual(albumData.length);
    expect(
      component.queryByRole('table', { name: /search results/i })
    ).toBeNull();

    const searchField = component.getByRole('textbox', {
      name: /search field/i,
    });

    fireEvent.change(searchField, { target: { value: 'search' } });

    const searchResultsTable = await waitFor(() =>
      component.getByRole('table', { name: /search results/i })
    );
    const searchResults = await waitFor(() =>
      within(searchResultsTable).getAllByRole('row')
    );

    expect(searchResults.length).toEqual(trackData.length);
    expect(within(albumSection).getAllByRole('listitem').length).toEqual(4);

    // switch to mobile
    setScreenSize(400);
    component.rerender(<Home />);

    expect(component.queryByRole('list', { name: /album list/i })).toBeNull();
  });
});
