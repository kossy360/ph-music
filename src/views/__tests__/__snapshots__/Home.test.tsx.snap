// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search result view should render table correctly 1`] = `
<div
  aria-label="album list"
  class="albumsGrid"
  role="list"
/>
`;
